pkg/resource/deploy/deployment_executor.go

```diff
@@ -188,7 +188,23 @@ func (ex *deploymentExecutor) Execute(callerCtx context.Context) (*Plan, error)
 	if err != nil {
 		return nil, err
 	}
-	defer src.Close()
+	defer func() {
+		// If the context was canceled, we want to return quickly here and not wait for the program to complete.
+		wasCanceled := errors.Is(callerCtx.Err(), context.Canceled)
+		if wasCanceled {
+			logging.V(4).Infof("deploymentExecutor.Execute(...): context was canceled, not waiting for program completion")
+			return
+		}
+		closeErr := src.Close()
```

Comment:
A little suspect of that? Does that mean we can leave a runtime plugin and program running disconnected from the CLI? Should we not be sigkill-ing the program on context cancellation, so it should be fine to wait for it?

Reply:
If we try to `src.Close()` on cancel here, we will wait for the language runtime to exit on its own. This should happen eventually as all the in flight steps are canceled, but if there's for example a `time.sleep` or any other non-pulumi operations that take time, we wait for those too. For example I ran into this with the program in test TestRunCanceled, which has a `time.Sleep(10)` and we wait for that, where we previously didn't.

If we just return without closing, we cancel swiftly. I am not sure where we do terminate the plugin, but it doesn't otherwise stay running. Until very recently we never called `Close()`, so in that sense the behaviour is unchanged from what it was a week or two ago.